keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+> Note: if an object store supports more than one protocol then the admin should create a `BucketClass` per protocol.
+4. `anonymousAccessMode`: (optional) a string specifying *uncredentialed* access to the backend bucket.  This is applicable for cases where the backend storage is intended to be publicly readable and/or writable. One of:
+   - "private": Default, disallow uncredentialed access to the backend storage.
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+5. `retentionPolicy`: defines bucket retention for greenfield `BucketRequest`.
+   - _Retain_: (default) the `Bucket` instance and the backend bucket are preserved. The `Bucket` may potentially be reused.
+   - _Delete_: the backend bucket and the `Bucket` instance are deleted.
+> Note: the `Bucket`'s retention policy is set to "Retain" as a default. Exercise caution when using the "Delete" retention policy as the bucket content will be deleted.
+> Note: a `Bucket` is not deleted if it is bound to a `BucketRequest` or `BucketAccess`.
+6. `allowedNamespaces`: a list of namespaces that are permitted to either create new buckets or to access existing buckets. This list is static for the life of the `BucketClass`, but the `Bucket` instance's list of allowed namespaces can be mutated by the admin.
+7. `parameters`: (optional) a map of string:string key values.  Allows admins to set provisioner key-values.
+
+### Access APIs
+
+The Access APIs abstract the backend policy system.  Access policy and user identities are an integral part of most object stores.  As such, a system must be implemented to manage both user/credential creation and the binding of those users to individual buckets via policies.  Object stores differ from file and block storage in how they manage users, with cloud providers typically integrating with an IAM platform.  This API includes support for cloud platform identity integration with Kubernetes ServiceAccounts.  On-prem solutions usually provide their own user management systems, which may look very different from each other and from IAM platforms.  We also account for third party authentication solutions that may be integrated with an on-prem service.
+
+#### BucketAccessRequest
+
+A user namespaced custom resource representing an object store user and an access policy defining the user’s relation to that storage.  A user creates a `BucketAccessRequest` (BAR) in the app's namespace (which is the same namespace as the `BucketRequest`) A 'BucketAccessRequest' specifies *either* a ServiceAccount or a secret name which contains access policy.  Specifying a ServiceAccount enables provisioners to support cloud provider identity integration with their respective Kubernetes offerings.
```

Comment:
FIX: secret name is no longer part of the BAR.

Reply:
Just to clarify, do that means `serviceAccountName` becomes required now. If possible can u please explain how SA can be connected to user in storage provider, also can single service account can attach multiple BARs?

Reply:
I have the same concern regarding sharing service account across multiple BARs. I don't see a good mechanism to disallow that if we decide to do so.
If we share a service account there might be edge-cases related to granting/revoking access. For example if we share a service account across BARs with intersecting permissions, how we can properly revoke these permissions when a user deletes one of the BARs?

Reply:
@oleksiys This is something we went over. Conflicts can arise and we decided that if two different access policies are requested for the same SA, then we reject the second request.

Reply:
We'll update the KEP to reflect this. This was discussed in the review session: https://docs.google.com/presentation/d/180c1-fcRvGUU_AqMbe5swsDAbNg2yvLdQYdxnXVP9Xs/edit#slide=id.g907055d5e0_0_87

Reply:
@thotz wrote:
> Just to clarify, do that means serviceAccountName becomes required now. If possible can u please explain how SA can be connected to user in storage provider, also can single service account can attach multiple BARs?

`BAR.serviceAccountName` is still optional. If the name is provided then it will be used by the backend object store to map the SA to a user identity. If SA is omitted then the backend will mint a `principal` who is the user for this bucket.